### PR TITLE
crimson/admin: deal with seastar's new metric types/logics

### PR DESCRIPTION
**Pull request by xxhdx1985126**
Fixes: https://tracker.ceph.com/issues/55666
Signed-off-by: Xuehan Xu <xxhdx1985126@gmail.com>

## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/master/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "pacific"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

## Checklist
- Tracker (select at least one)
  - [ ] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [ ] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [ ] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [ ] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
</details>

### CODE REVIEW

File: src/crimson/admin/osd_admin.cc [hunk 1]
```diff
@@ -245,8 +245,11 @@ class DumpMetricsHook : public AdminSocketHook {
     case data_type::GAUGE:
       f->dump_float(value_name, v.d());
       break;
+    case data_type::REAL_COUNTER:
+      f->dump_float(value_name, v.d());
+      break;
     case data_type::COUNTER:
-      f->dump_unsigned(value_name, v.ui());
+      f->dump_int(value_name, v.i());
```
Comment (tchaikov) [inline src/crimson/admin/osd_admin.cc:252]: `COUNTER` should be considered as unsigned, i think.

Comment (xxhdx1985126) [reply src/crimson/admin/osd_admin.cc:252]: Um, the most recent version of seastar seems to allow negative counter: https://github.com/scylladb/seastar/blob/693dac6c030e02cab3ab87a4c818044d935837cc/src/core/reactor.cc#L2279

Comment (tchaikov) [reply src/crimson/admin/osd_admin.cc:252]: i see. seastar does not enforce the signness of this type, and it stores it using a `double`.  and it considers it as an unsigned when exporting it to collected, see https://github.com/scylladb/seastar/blob/693dac6c030e02cab3ab87a4c818044d935837cc/src/core/scollectd.cc#L217-L220

but considers as a signed when exporting it to prometheus, see https://github.com/scylladb/seastar/blob/693dac6c030e02cab3ab87a4c818044d935837cc/src/core/prometheus.cc#L63-L64 . but according to prometheus' doc (https://prometheus.io/docs/concepts/metric_types/), a counter can only be unsigned. so i think unsigned is a better choice. also, an unsigned integer is able to represent a greater value.

Comment (xxhdx1985126) [reply src/crimson/admin/osd_admin.cc:252]: > i see. seastar does not enforce the signness of this type, and it stores it using a `double`. and it considers it as an unsigned when exporting it to collected, see https://github.com/scylladb/seastar/blob/693dac6c030e02cab3ab87a4c818044d935837cc/src/core/scollectd.cc#L217-L220
> 
> but considers as a signed when exporting it to prometheus, see https://github.com/scylladb/seastar/blob/693dac6c030e02cab3ab87a4c818044d935837cc/src/core/prometheus.cc#L63-L64 . but according to prometheus' doc (https://prometheus.io/docs/concepts/metric_types/), a counter can only be unsigned. so i think unsigned is a better choice. also, an unsigned integer is able to represent a greater value.

Yes, I agree. But seastar's metric `cpu_steal_time_ms` being negative is causing admin `dump_metrics` command's failure, right now. Should we drop a patch to the seastar project?

Comment (tchaikov) [reply src/crimson/admin/osd_admin.cc:252]: i'd do so if i have had an explanation and a sound solution to this issue. to my best understanding, `reactor::total_steal_time()` should return a duration in nanoseconds. by reading the comment of that member function, i think this value should be positive. 

do you have a fix? if yes, would be better to upstream it.

Comment (xxhdx1985126) [reply src/crimson/admin/osd_admin.cc:252]: Um, not yet. I actually don't quite understand why that value can be negative......

**Comment by xxhdx1985126** [comment]
@tchaikov I changed the code to treat negative counter values as zero, please take a look, thanks:-)

**Review by cyx1231st — APPROVED** [review]
I also got the same issue. Approve as the fix works for me.

**Review by tchaikov — APPROVED** [review]

**Pull request merged**